### PR TITLE
Plant continuous integration

### PR DESCRIPTION
**Pull request by lucafavatella**
Please refer to commit messages for details.

### CODE REVIEW

**Comment by charlag** [comment]
Hi! Thanks for your PR
I've noticed this PR now and I talked to the team. We make internal builds for the client but we would also be interested in the external build system as well.
I have some experience with CircleCI (they give you 4 parallel machines) but I think Travis is perfectly fine too.
Are you still interested in working on the CI?

File: test/test.js [hunk 1]
```diff
@@ -52,8 +52,16 @@ function runTest() {
...
 		let testRunner = child_process.fork(`../build/test/${project}/bootstrapNode.js`)
-		testRunner.on('exit', () => {
+		console.log(`> spawned child process ${testRunner.pid}`)
+		testRunner.on('error', (err) => {
+			console.log(`> child process errored: ${err}`);
+			process.exit(1)
+		})
+		testRunner.on('exit', (code) => {
+			console.log(`> child process exited with code ${code}`);
+			if (code !== 0) process.exit(1)
```
Comment (lucafavatella) [inline test/test.js]: I see this handled - better - in [merged code](https://github.com/tutao/tutanota/commit/edb1e9346f0e88ca85ecaecabfc45dd5093984f5#diff-c1129c8b045390789fa8ff62f2c6b4a9R61).

File: test/test.js [hunk 2]
```diff
@@ -52,8 +52,16 @@ function runTest() {
 	if (testRunner != null) {
 		console.log("> skipping test run as test are already executed")
 	} else {
+		console.log("> spawn child process")
 		let testRunner = child_process.fork(`../build/test/${project}/bootstrapNode.js`)
-		testRunner.on('exit', () => {
+		console.log(`> spawned child process ${testRunner.pid}`)
+		testRunner.on('error', (err) => {
```
Comment (lucafavatella) [inline test/test.js]: This case seems still not handled in master code. Not sure if this can happen.

File: package.json [hunk 3]
```diff
@@ -2,6 +2,13 @@
   "name": "tutanota",
   "version": "3.25.11",
   "scripts": {
+    "pretest": "npm run-script flow",
+    "test": "npm run-script testapi && npm run-script testclient",
```
Comment (lucafavatella) [inline package.json]: I see something similar to this included in [merged code](https://github.com/tutao/tutanota/commit/0708dae2#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R12).

File: package.json [hunk 4]
```diff
@@ -2,6 +2,13 @@
...
+    "pretestapi": "mkdir -p build/src/api/worker && test/api.sh server spawn \"$(pwd)/server.js\"",
+    "testapi": "( cd test && node test.js api; )",
```
Comment (lucafavatella) [inline package.json]: I see something similar to this included in [merged code](https://github.com/tutao/tutanota/commit/0708dae2#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R12), though not sure how that code handles:
- Potential bad interactions with other `server.js` instances.
- Distinct exit codes.

(See test/api.sh in this PR for details.)

File: package.json [hunk 5]
```diff
@@ -1,6 +1,9 @@
 {
   "name": "tutanota",
   "version": "3.25.11",
+  "directories": {
+    "test": "./test"
```
Comment (lucafavatella) [inline package.json]: This seems not included in master though seems only doc so not relevant.

File: .travis.yml [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+script:
+  - |
+    case "${JOB:?}" in
+      test-server-api) npm install && npm run-script testapi ;;
+      test-web-client) npm install && npm run-script testclient ;;
+      npm-test) npm install-test ;;
+    esac
+after_failure:
+  - tail -n 100 npm-debug.log
+env:
```
Comment (lucafavatella) [inline .travis.yml]: This commit is simply hooking the tests to CI - with environment setup. I see no public CI in master.

**Comment by lucafavatella** [comment]
It looks like most test improvements in this PR (propagation of exit code in tests, and scripts for running test tasks) are now included in master. That shall help future contributors - thanks.

> Are you still interested in working on the CI?

No.

**Pull request closed without merging**